presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -211,7 +253,62 @@ public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph
                     joinFilters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(joinFilters)),
                     Optional.empty(),
                     Optional.empty(),
-                    Optional.empty());
+                    Optional.empty()));
+        }
+
+        private boolean introducesCrossJoin(List<PlanNode> leftSources, List<PlanNode> rightSources)
+        {
+            List<PlanNode> allSources = ImmutableList.<PlanNode>builder()
+                    .addAll(leftSources)
+                    .addAll(rightSources)
+                    .build();
+            Set<PlanNode> leftSet = ImmutableSet.copyOf(leftSources);
+            Set<PlanNode> rightSet = ImmutableSet.copyOf(rightSources);
+            Set<Set<PlanNode>> connectedNodeSets = getConnectedNodeSets(allSources);
+
+            for (Set<PlanNode> connectedSet : connectedNodeSets) {
```

Comment:
Yes, but checking for `joinConditions` emptiness when building actual node would ensure that, right?